File: exercise/LeetCode/763. Partition Labels/Partition Labels.py

class Solution:
    def partitionLabels(self, S):
        """
        :type S: str
        :rtype: List[int]
        """

        """
        # 容易理解，但速度太慢
        num = []
        while S:
            k = 1
            # 两个集合没有交集，即可分开
            while not set(S[:k]).isdisjoint(set(S[k:])):
                k += 1
            num.append(k)
            S = S[k:]
        return num
        """

        p = []
        for k in range(len(S)):
            # p中无含有S[k]的字符串
            flg = False
            index = -1
            for i in range(len(p)):
                if S[k] in p[i]:
                    # p中有含有S[k]的字符串
                    index = i
                    flg = True
                    break
            p.append(S[k])
            if flg:
                # 合并index到len(p)的字符串
                for i in range(index+1, len(p)):
                    p[index] += p[i]
                p = p[0:index+1]
        num = []
        for k in p:
            num.append(len(k))
        return num


if __name__ == "__main__":
    su = Solution()
    print(su.partitionLabels("ababcbacadefegdehijhklij"))

File: exercise/LeetCode/763. Partition Labels/Partition Labels.cpp

#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    vector<int> partitionLabels(string S) {
        std::vector<int> vi;
#if(1)
        // 记录每个字母的最大的下标
        std::vector<int> pos(26, 0);
        for (unsigned int k = 0; k < S.size(); ++ k)
            pos[S[k]-'a'] = k;
        int is=0, ie=-1;
        for (int k = 0; k < S.size(); ++ k)
        {
            if (pos[S[k]-'a'] > ie) ie = pos[S[k]-'a'];
            if (k == ie)
            {
                vi.push_back(ie-is+1);
                is = k+1;
            }
        }
#else
        std::vector<std::string> vs;
        for (char c:S)
        {
            bool flg = false;
            unsigned int i;
            for (i = 0; i < vs.size(); i ++)
            {
                if (std::string::npos != vs[i].find(c))
                {
                    flg = true;
                    break;
                }
            }
            vs.push_back(std::string(""));
            vs[vs.size()-1] += c;
            if (flg)
            {
                int size = vs.size();
                for (int j = i + 1; j < size; j ++)
                    vs[i] += vs[j];
                for (int j = i + 1; j < size; j ++)
                    vs.pop_back();
            }
        }
        for (auto it:vs)
            vi.push_back(it.size());
#endif
        return vi;
    }
};

int main(void)
{
    Solution su;
    std::vector<int> vi = su.partitionLabels("ababcbacadefegdehijhklij");
    for (auto k:vi)
        std::cout << k << std::endl;
    return 0;
}
